clippy_lints/src/methods/search_is_some.rs

```diff
@@ -142,3 +156,12 @@ pub(super) fn check<'tcx>(
         }
     }
 }
+
+fn is_receiver_of_method_call(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
```

Comment:
nit: This should go in `utils/` somewhere (and get some documentation)


And it's possible we already have something like this in utils?

Reply:
Shouldn't this also actually check that the receiver hir id matches `expr.hir_id`? Otherwise this can return true for something like `foo.bar(expr)` (the parent of `expr` is a method call because it's an argument, but `expr` is not the receiver).

(Also I wonder if this sort of check would be better done by checking if `expr.precedence() > PREC_PREFIX` ? Method calls aren't the only expressions where parentheses are necessary, I think. Field and index expressions also have higher precedence than `!`)

Reply:
Yeah, I hadn't even looked at the implementation.

Checking precedence sounds great!

Reply:
> Shouldn't this also actually check that the receiver hir id matches `expr.hir_id`? Otherwise this can return true for something like `foo.bar(expr)` (the parent of `expr` is a method call because it's an argument, but `expr` is not the receiver).
> 
> (Also I wonder if this sort of check would be better done by checking if `expr.precedence() > PREC_PREFIX` ? Method calls aren't the only expressions where parentheses are necessary, I think. Field and index expressions also have higher precedence than `!`)

Thanks for the heads up! I've updated the code to check that the receiver's hir_id matches the expr's hir_id.

Reply:
Regarding checking the precedence, I have a concern in this case because if we directly check `expr.precedence().order() > PREC_PREFIX` it will always return `true` because `expr` is a `MethodCall` and therefore has `ExprPrecedence::. MethodCall` precedence, which is greater than `PREC_PREFIX`. This means that we will always add a parenthesis to the generated suggestions.

I wonder if it would be better to just create separate cases for these (field access, index access, method call) when checking parent expression.

Reply:
> nit: This should go in `utils/` somewhere (and get some documentation)
> 
> And it's possible we already have something like this in utils? (cc @Alexendoo @blyxyas who might remember)

Sounds good! I can put it in `utils` after we've come to a conclusion of how to implement it.

Reply:
> Regarding checking the precedence, I have a concern in this case because if we directly check `expr.precedence().order() > PREC_PREFIX` it will always return `true` because `expr` is a `MethodCall` and therefore has `ExprPrecedence::. MethodCall` precedence, which is greater than `PREC_PREFIX`. This means that we will always add a parenthesis to the generated suggestions.

I meant checking the precedence of the parent expression, which shouldn't always be a method call. For `...find().is_none() && true` the parent is the `&&` expression with a precedence lower than `!` which wouldn't need parentheses (or there might not be a parent expression at all, in which case we also don't need parentheses)

> I wonder if it would be better to just create separate cases for these (field access, index access, method call) when checking parent expression.

On second thought, it probably might be fine. I realized that field and index expressions don't really make sense in this context: there's no way to index into a bool or access any fields, and a bool can't be called either. I think the only type of expression with a higher precedence than `!` that makes sense in this context are method calls.

Reply:
> I meant checking the precedence of the parent expression, which shouldn't always be a method call. For `...find().is_none() && true` the parent is the `&&` expression with a precedence lower than `!` which wouldn't need parentheses (or there might not be a parent expression at all, in which case we also don't need parentheses)

In the case of `foo.bar(expr)`, it doesn't seem to work. As you mentioned, the parent expression of `expr` is a `MethodCall`, which has a higher precedence than `PREC_PREFIX`. Therefore, it seems that we still need to create a separate case for this.

> On second thought, it probably might be fine as is. I realized that field and index expressions don't really make sense in this context: there's no way to index into a bool or access any fields, and a bool can't be called either. I think the only type of expression with a higher precedence than `!` that makes sense in this context are method calls, so the current way of checking it would already cover all cases

Yes, I was thinking about this yesterday. But since Manishearth mentioned yesterday that this function should be in `utils/`, I was thinking that maybe we should implement a generic version of this method.

Reply:
If we move it into utils, I agree that it should be more generic.

Since the PR has been open so long, I would be okay with merging this with a simpler version and leaving the generic version as a future TODO